Rework Get Item Table to be more flexible for adding custom items

leggettc18 opened the pull request:
This relocates sGetItemTable from `z_player.c` into a more centralized location, and into a more flexible format for mods to be able to add their own custom item tables. It also moves randomizer's custom items out of the main item table and removes their enum values from `GetItemID` and `ItemID`. The randomizer items are instead indexed by their `RandomizerGet` enum values. The randomizer Item Table was tricky to implement, because some of the items have vanilla `ItemIDs` but not vanilla `GetItemIDs`. This means they needed a custom `GetItem` entry, but not custom `Item_Give`  code like all the others. Hence storing the `ModIndex` enum both in `OTRGlobals::Instance` and as a property of the `GetItemEntry` struct. The one in `OTRGlobals` is for determining which table to use, and the one in the `GetItemEntry` struct is for determining whether to run `Item_Give` or `Randomizer_Item_Give`.

This still needs cleanup but I would appreciate some general architectural review. There is one bug I'm aware of where if an item from the rando table and an item from the vanilla table are both trying to be given to link at once, one ModIndex will win out over the other and Link will either receive the wrong item or possibly crash the game. An example of this is in the Temple of Time. If you set LACS to be Bottle With Bugs (Rando Table) and Prelude to Deku Nuts (5) (Vanilla Table), the LACS check will become a Heart Container, since it's taking `RG_BOTTLE_WITH_BUGS` and using it as the `getItemID` for the Vanilla Table. The reason for this is that both attempts to give the check are running at the same time, and the way I'm setting which table to use doesn't work very well in that scenario. The ModIndex gets set back to `MOD_VANILLA` after starting the item get code but before the LACS reward actually gets pulled out of the Item Table. I can think of a few ways to fix it, such as returning more than just an ItemID from the rando item id obtaining code and/or setting extra properties on the player struct, but they are all going to require a *lot* of changes to existing actor code that I was trying to avoid changing.

leggettc18 committed:
Adds ItemTableManager class.

leggettc18 committed:
Implements new getItem table in game.

leggettc18 committed:
Adds rando item table and way to differentiate tables in GetItemEntry.

leggettc18 committed:
Resolve merge conflicts with develop-zhora

leggettc18 committed:
Adds rough ability to differentiate between rando and vanilla items.merge stashed changes from before develop-zhora merge

leggettc18 committed:
Change ItemTableID to be uint16_t so we can use ModIndex for it.

aMannus commented:
First of all, you're doing god's work, holy crap.

Something that might be food for thought: How would we handle if rando treats a certain vanilla item differently from vanilla? For example, let's say we add "Chest matches size contents" down the line, which might be related to the short or long animation of opening a chest. But then only major items should be a big chest, and items like pieces of heart should be a small chest. But a piece of heart has the long animation in the vanilla item table.

Or as we had problems with before, ice traps having a short animation in vanilla, but maybe we actually want them to have a long animation once the option to have ice traps appear as major items is in.

leggettc18 commented:
> Something that might be food for thought: How would we handle if rando treats a certain vanilla item differently from vanilla? For example, let's say we add "Chest matches size contents" down the line, which might be related to the short or long animation of opening a chest. But then only major items should be a big chest, and items like pieces of heart should be a small chest. But a piece of heart has the long animation in the vanilla item table.

It should be easy enough to add a Chest Size property to the GetItemEntry struct I think. Can have it default to small or large depending on the chest animation but have specific overrides where necessary, probably by having the existing macro for `GET_ITEM` set the default one and add a new one that allows specifying a chest size.

> Or as we had problems with before, ice traps having a short animation in vanilla, but maybe we actually want them to have a long animation once the option to have ice traps appear as major items is in.

For Ice Traps, part of this is that vanilla Ice Traps at least will behave exactly as they did in Vanilla, since when we're not in a rando save we'll just have the same getItemEntry that vanilla was using. It's not necessarily coded this way right this minute but rando will have it's own Ice Trap item entry which might have the fool textbox associated with it for situations where that textID might be required.

aMannus commented:
Alright, sounds good to me!

leggettc18 commented:
If I wanted to treat a vanilla item differently in rando, I would probably create a GetItemEntry with the vanilla ItemID, whatever object, drawID, and chest animation I wanted, and set the ModIndex to `MOD_RANDOMIZER`. Then use `Randomizer_Item_Give` to determine how to specially handle it if necessary, or I can intercept it somewhere else in `z_player.c` if absolutely necessary.

leggettc18 committed:
Should fix switch build

leggettc18 committed:
Should fix switch build pt 2

leggettc18 committed:
merge in develop-zhora

leggettc18 committed:
Adds new files to CMakeLists.

leggettc18 committed:
Implements fixes for competing getItem calls.

leggettc18 committed:
Correctly renders freestanding items

Particle effects are probably broken, need to fix them still, I pretty
much know how I would do that.

leggettc18 commented:
Fixed the LACS/Prelude competing issue, and can fix any others that crop up in testing. Also fixed freestanding item models. Custom Particles are most likely broken, can fix them just will need some time.

leggettc18 committed:
Fixed Particle effects in the new getItem system.

leggettc18 committed:
Fixes item fanfares

leggettc18 committed:
Partially fixes Ice Traps

Obtaining a freestanding Ice Trap causes link to slide forward
and receiving one from an NPC plays the sound effect and damage
animation but doesn't freeze link.

leggettc18 committed:
Merge branch 'develop-zhora' of https://github.com/HarbourMasters/Shipwright into get-item-rework

leggettc18 commented:
Ok, as of now the only things that aren't working correctly are Ice Traps in certain situations. They don't crash the game or anything but Link keeps sliding forward from a freestanding ice trap and when given by an NPC he plays the damage animation and ice sound effect but does not freeze.

leggettc18 committed:
Some more partial ice trap fixing that wasn't pushed earlier

leggettc18 commented:
I found out that Ice Traps weren't broken by anything in this commit, they've been broken for about a week. A fix for that is in #1090 

With that, I think this is ready for a more in-depth review. I know of a few minor relocations and renamings that I want to do but any reviewers can feel free to chime in on anything they feel like.

briaguya0 commented on soh/soh/Enhancements/item-tables/ItemTableManager.cpp:
```diff
@@ -0,0 +1,59 @@
+    return itemTables.emplace(tableID, newItemTable).second;
+}
+
+bool ItemTableManager::AddItemEntry(uint16_t tableID, uint16_t getItemID, uint16_t itemID, uint16_t objectID, int16_t drawID, uint16_t textID, uint16_t field, bool chestAnim) {
+    ItemTable* itemTable = RetrieveItemTable(tableID);
+    if (itemTable == nullptr) {
+        return false;
+    }
+    GetItemEntry getItemEntry = GET_ITEM(itemID, objectID, drawID, textID, field, chestAnim);
+    return itemTable->emplace(getItemID, getItemEntry).second;
+}
+
+bool ItemTableManager::AddItemEntry(uint16_t tableID, uint16_t getItemID, GetItemEntry getItemEntry) {
+    ItemTable* itemTable = RetrieveItemTable(tableID);
+    if (itemTable == nullptr) {
+        return false;
+    }
+    return itemTable->emplace(getItemID, getItemEntry).second;
+}
```
is there a way to handle this with less copypasta? i feel like it'd be nice to have the top one just create the GetItemEntry and then call the bottom one

leggettc18 replied:
Just wanted to have a way to do it that didn't require GET_ITEM in case there were situations where it was difficult to use for some reason. But I'm not using said function anywhere so I could remove it.

briaguya0 commented on soh/soh/Enhancements/item-tables/ItemTableManager.cpp:
```diff
@@ -0,0 +1,59 @@
+GetItemEntry ItemTableManager::RetrieveItemEntry(uint16_t tableID, uint16_t itemID) {
+    ItemTable* itemTable = RetrieveItemTable(tableID);
+    if (itemTable != nullptr) {
+        auto foundItemEntry = itemTable->find(itemID);
```
nitpick (and totally your call on the naming) but i think just `itemEntry` would be nicer than `foundItemEntry` here

leggettc18 replied:
I ended up replacing all the calls to `::find` here with `::at` and added exception handling. Less questionable names and fewer loc.

briaguya0 commented on soh/soh/Enhancements/item-tables/ItemTableManager.cpp:
```diff
@@ -0,0 +1,59 @@
+    if (itemTable != nullptr) {
+        auto foundItemEntry = itemTable->find(itemID);
+        if (foundItemEntry != itemTable->end()) {
+            return foundItemEntry->second;
```
thinking about this now, it might be nice to come up with a different name than `itemEntry` or `foundItemEntry` since it's coming from the map, so we're getting both the `uint16_t` and the `GetItemEntry`

not sure what the best way to name this to make it clear what's going on is. it's fine as is, i just wish i had a better name in mind

leggettc18 replied:
Well the idea is that the `foundItemEntry` is used to verify whether or not we were able to find the `itemEntry` and by the time we assign anything to the `itemEntry` variable it is just the GetItemEntry and not the whole pair. In hindsight, it might be more readable to use `unordered_map::at` to get the itemEntry, and return GET_ITEM_NONE when the not found exception is caught instead.

leggettc18 replied:
And this would apply to custom_messages as well so I'd probably want to update that PR with the same thing if that sounds better to you.

briaguya0 replied:
hmm, i don't really have strong feelings one way or the other about the current method vs a try/catch. if the try/catch feels like it'll be more readable let's go for it! (it'll be fine to update the message stuff to use that pattern in this PR)

briaguya0 commented on soh/soh/Enhancements/item-tables/ItemTableManager.cpp:
```diff
@@ -0,0 +1,59 @@
+    if (foundItemTable == itemTables.end()) {
+        return nullptr;
+    }
+    ItemTable& itemTable = foundItemTable->second;
```
same thing as `foundItemEntry` here, same lack of ideas for a better name

briaguya0 commented on soh/soh/Enhancements/item-tables/ItemTableManager.h:
```diff
@@ -0,0 +1,33 @@
+#define CHEST_ANIM_LONG 1
+
+#define GET_ITEM(itemId, objectId, drawId, textId, field, chestAnim) \
+    { itemId, field, (chestAnim != CHEST_ANIM_SHORT ? 1 : -1) * (drawId + 1), textId, objectId }
```
i think it might be nice to check for `== CHEST_ANIM_LONG` instead, but it shouldn't matter

leggettc18 replied:
Pulled that straight from z_player.c, but I can change it.

briaguya0 replied:
yeah, i saw that it was that way already, for some reason i thought it was originally just `chestAnim ? 1 : -1` and was changed at some point but from the blame it seems it's been != short the entire time

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2028,8 +2029,8 @@ GetItemID Randomizer::GetItemFromGet(RandomizerGet randoGet, GetItemID ogItemId)
-        case RG_DOUBLE_DEFENSE:
-            return GI_DOUBLE_DEFENSE;
+        /*case RG_DOUBLE_DEFENSE:
+            return GI_DOUBLE_DEFENSE;*/
```
delete not comment out

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2225,7 +2226,7 @@ GetItemID Randomizer::GetItemFromGet(RandomizerGet randoGet, GetItemID ogItemId)
             return GI_NOCTURNE_OF_SHADOW;
         case RG_PRELUDE_OF_LIGHT:
-            return GI_PRELUDE_OF_LIGHT;
+            return GI_PRELUDE_OF_LIGHT;*/
```
delete not comment out

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2306,7 +2307,7 @@ GetItemID Randomizer::GetItemFromGet(RandomizerGet randoGet, GetItemID ogItemId)
             return GI_MEDALLION_SHADOW;
         case RG_LIGHT_MEDALLION:
-            return GI_MEDALLION_LIGHT;
+            return GI_MEDALLION_LIGHT;*/
```
delete not comment out

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2328,8 +2329,8 @@ GetItemID Randomizer::GetItemFromGet(RandomizerGet randoGet, GetItemID ogItemId)
-        case RG_ICE_TRAP:
-            return GI_ICE_TRAP;
+        // case RG_ICE_TRAP:
+        //     return GI_ICE_TRAP;
```
delete not comment out

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1526,14 +1665,62 @@ extern "C" int Randomizer_CopyHintFromCheck(RandomizerCheck check, char* buffer,
+}
+
 extern "C" s32 Randomizer_GetRandomizedItemId(GetItemID ogId, s16 actorId, s16 actorParams, s16 sceneNum) {
+    if (OTRGlobals::Instance->gRandomizer->CheckContainsRandoItem(OTRGlobals::Instance->gRandomizer->GetCheckFromActor(sceneNum, actorId, actorParams))) {
+        OTRGlobals::Instance->getItemModIndex = MOD_RANDOMIZER;
+    } else {
+        OTRGlobals::Instance->getItemModIndex = MOD_VANILLA;
+    }
```
where did we land on this working? having the mod index as a global thing wasn't related to the issue?

leggettc18 replied:
Right now there's a mix of this way and the fixed way, since this way works like 95% of the time and changing everything would take a while. Meanwhile people can use this for keysanity and such.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4727,8 +4778,139 @@ void DrawRandoEditor(bool& open) {
+    };
+    ItemTableManager::Instance->AddItemTable(MOD_RANDOMIZER);
+    for (int i = 0; i < ARRAY_SIZE(getItemTable); i++) {
+        if (i < 21) {
```
this feels a little magic numbery for my taste, maybe we could break this up into 2 arrays instead?

since the loop adds everything using
```
ItemTableManager::Instance->AddItemEntry(MOD_RANDOMIZER, getItemID, getItemTable[i]);
```

it should work to just loop through a `vanillaRandoGetItemTable` and a `randoExclusiveGetItemTable` or something

leggettc18 replied:
Yeah could do that, didn't think about it being two arrays for some reason.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizerTypes.h:
```diff
@@ -806,6 +806,8 @@ typedef enum {
     RG_PROGRESSIVE_STICK_UPGRADE,
     RG_PROGRESSIVE_BOMBCHUS,
     RG_PROGRESSIVE_MAGIC_METER,
+    RG_MAGIC_SINGLE, // Added for refactor of GetItemEntries
+    RG_MAGIC_DOUBLE, // Added for refactor of GetItemEntries
```
i'm assuming we need these because we're using `RG`s as `getItemID`s now? might be nice to update the comments to be a little more descriptive

leggettc18 replied:
That's right.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.h:
```diff
@@ -6,6 +6,10 @@
 
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
+
+typedef enum { MOD_VANILLA, MOD_RANDOMIZER } ModIndex;
```
not the biggest fan of the naming for `MOD_VANILLA`

not sure what a better name would be though

leggettc18 replied:
Yeah my thought was that later on registered mods would get an index starting at 2 when they register themselves, vanilla would always be 0 and rando would always be 1. Hence they have the enums for readability. And of course changing this to use some other identifier than a number altogether is also an option.

briaguya0 replied:
i guess i'm cool with the name `MOD_VANILLA` if i just think of mod as short for module instead of modification

leggettc18 replied:
Could also change to `MOD_NONE` now that I think about it, would that be better?

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.h:
```diff
@@ -33,14 +39,15 @@ class Randomizer {
     std::string GetHintFromCheck(RandomizerCheck check);
     GetItemID GetRandomizedItemIdFromKnownCheck(RandomizerCheck randomizerCheck, GetItemID ogId);
     GetItemID GetRandomizedItemId(GetItemID ogId, s16 actorId, s16 actorParams, s16 sceneNum);
+    bool CheckContainsRandoItem(RandomizerCheck randoCheck);
```
might be nice to name this `CheckContainsRandoExclusiveItem` or something, but that might be confusing because of the randomizer items that use vanilla itemids vs the randomizer items that are all new, names are hard

leggettc18 replied:
Yeah that's actually exactly why I didn't put the word Exclusive in it.

leggettc18 replied:
The rando exclusive items have their modIndex in their ItemEntry so the game can differentiate that at Item_Give time for the rando exclusive items. This is what the code uses to determine which table to pull from, but still may pull a vanilla item from the rando table, which didn't have a vanilla getItemEntry because it was `Item_Give`en manually.

briaguya0 replied:
hmm, what if it was `CheckContainsVanillaItem` and then the places where it's used to see if something should use the rando table or not we can use `!` (unless you foresee issues with that with future mods)

leggettc18 replied:
Most likely randomizer is the only one that will use this, most other mods will probably just call ItemTable_RetrieveEntry or something like that since they won't have the same check return different items in different situations, and if they do they will probably have their own logic for determining that. Inverting that function to check for Vanilla Items may be better then, so we don't have to update that function when implementing any new `RG_` values down the line.

briaguya0 replied:
> so we don't have to update that function when implementing any new RG_ values down the line.

hadn't thought about needing to update the function, but yeah, good call. inverting it seems like a good plan

leggettc18 replied:
Addressed by 2077636c920ccd59c0ff864a7cbab28682e30467

briaguya0 commented on soh/src/code/z_en_item00.c:
```diff
@@ -1208,38 +1208,56 @@ void EnItem00_Draw(Actor* thisx, GlobalContext* globalCtx) {
-        case GI_STICK_UPGRADE_30:
-            color_slot = 6;
+    switch (giEntry.modIndex) {
+        case 0:
```
magic number, probably should use the enum here

briaguya0 commented on soh/src/code/z_en_item00.c:
```diff
@@ -1208,38 +1208,56 @@ void EnItem00_Draw(Actor* thisx, GlobalContext* globalCtx) {
-        case GI_NUT_UPGRADE_30:
-        case GI_NUT_UPGRADE_40:
-            color_slot = 7;
+        case 1:
```
magic number use enum

briaguya0 commented on soh/src/code/z_parameter.c:
```diff
@@ -2237,6 +2173,84 @@ u8 Item_Give(GlobalContext* globalCtx, u8 item) {
+    }
+
+    temp = gSaveContext.inventory.items[slot];
+    osSyncPrintf("Item_Register(%d)=%d  %d\n", slot, item, temp);
```
`Item_Register`?

leggettc18 replied:
I think that came from Vanilla Item_Give that I copied over to here to mimic the structure of that in the Rando version. Registering an Item in Link's Inventory is probably what they mean.

briaguya0 commented on soh/src/code/z_sram.c:
```diff
@@ -367,43 +367,43 @@ void GiveLinkSong(GetItemID getItemId) {
+void GiveLinkDungeonReward(uint16_t getItemId) {
     s16 item;
 
     u8 medallion = 0;
```
i know this is unrelated to your changes, just noticed this and thought it'd be good to change this to a bool

briaguya0 replied:
made issue https://github.com/HarbourMasters/Shipwright/issues/1157

briaguya0 commented on soh/src/code/z_sram.c:
```diff
@@ -367,43 +367,43 @@ void GiveLinkSong(GetItemID getItemId) {
-        case GI_MEDALLION_FOREST:
+        case RG_FOREST_MEDALLION:
             item = ITEM_MEDALLION_FOREST;
             medallion = 1;
```
and change these to be `= true`

briaguya0 replied:
https://github.com/HarbourMasters/Shipwright/issues/1157

briaguya0 commented on soh/src/code/z_sram.c:
```diff
@@ -367,43 +367,43 @@ void GiveLinkSong(GetItemID getItemId) {
+        case RG_ZORA_SAPPHIRE:
             item = ITEM_ZORA_SAPPHIRE;
             break;
     }
```
and just be able to say `if (medallion)` on line 411

briaguya0 replied:
https://github.com/HarbourMasters/Shipwright/issues/1157

briaguya0 commented on soh/src/overlays/actors/ovl_En_Box/z_en_box.c:
```diff
@@ -70,6 +70,7 @@ static InitChainEntry sInitChain[] = {
 };
 
 static UNK_TYPE sUnused;
+int32_t sItem;
```
`sItem`?

leggettc18 replied:
Follows the naming convention of other static variable names in existing actors.

briaguya0 commented on soh/src/overlays/actors/ovl_En_Xc/z_en_xc.c:
```diff
@@ -296,6 +296,7 @@ void GivePlayerRandoRewardSheikSong(EnXc* sheik, GlobalContext* globalCtx, Rando
         GetItemID getItemId = Randomizer_GetItemIdFromKnownCheck(check, ogSongId);
         if (check == RC_SHEIK_AT_TEMPLE && !Flags_GetTreasure(globalCtx, 0x1F)) {
             if (func_8002F434(&sheik->actor, globalCtx, getItemId, 10000.0f, 100.0f)) {
+                player->getItemEntry = Randomizer_GetItemFromKnownCheck(check, ogSongId);
```
we already have the `getItemId` from line 296 here, is there a reason to call into this again?

leggettc18 replied:
That's not getting the getItemId, that's getting the getItemEntry. I'm not finished with the entire conversion yet, but eventually the idea is that we only need to call this and not the one for the getItemId. That won't change how people register items to the table though or how they apply says items to Link's inventory, which is why I wanted to get this PR out there, since we've got people making keysanity and other things which would benefit greatly from this.

briaguya0 replied:
ah, missed that (names are very similar) so if we changed line 296 here to get the entry, then replaced getItemId in line 298 with the id from the entry, we could avoid the extra call

leggettc18 replied:
Eventually yeah, but I think some things in z_player.c still depend on having player->getItemId so I'd need to change those first. I could probably add the getItemId to the GetItemEntry struct though as a way to only have one call. Then the situations where we don't need the getItemId just won't use it.

briaguya0 replied:
yeah, forgot the entry didn't include the getItemId for a second, the itemid vs getitemid stuff strikes again! this is fine as is, but yeah, maybe adding the getitemid to the struct could be a cool future thing

leggettc18 replied:
It'd be easy enough to do I might as well cut out the extra calls while I'm fixing the other things you mentioned. I'll be able to get on it soon now that custom messages is merged.

leggettc18 replied:
Addressed in 11db45eef6d6c7cbd2b438de5a69665ce0544738

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -21,13 +21,13 @@
-    /* 0x03 */ u8 textId;
-    /* 0x04 */ u16 objectId;
-} GetItemEntry; // size = 0x06
+//typedef struct {
+//    /* 0x00 */ u8 itemId;
+//    /* 0x01 */ u8 field; // various bit-packed data
+//    /* 0x02 */ s8 gi;    // defines the draw id and chest opening animation
+//    /* 0x03 */ u8 textId;
+//    /* 0x04 */ u16 objectId;
+//} GetItemEntry; // size = 0x06
```
delete not comment

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12715,30 +12738,69 @@ s32 func_8084DFF4(GlobalContext* globalCtx, Player* this) {
-            (this->getItemId == GI_HEART)) {
-            Audio_PlaySoundGeneral(NA_SE_SY_GET_BOXITEM, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
+        Message_StartTextbox(globalCtx, giEntry.textId, &this->actor);
+        if (giEntry.modIndex == 0) {
```
magic number

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12715,30 +12738,69 @@ s32 func_8084DFF4(GlobalContext* globalCtx, Player* this) {
+        Player_SetPendingFlag(this, globalCtx);
+
+        if (this->getItemEntry.objectId != OBJECT_INVALID) {
+            if (giEntry.modIndex == 0) {
```
magic number

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12715,30 +12738,69 @@ s32 func_8084DFF4(GlobalContext* globalCtx, Player* this) {
+                    }
+                    Audio_PlayFanfare(temp1);
+                }
+            } else {
```
should probably else if here

leggettc18 replied:
Addressed by 0c334615b65abe06a7d7e9a3ceb373c6e28e7203

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12715,30 +12738,69 @@ s32 func_8084DFF4(GlobalContext* globalCtx, Player* this) {
-            else {
-                temp1 = temp2 = (this->getItemId == GI_HEART_PIECE) ? NA_BGM_SMALL_ITEM_GET : NA_BGM_ITEM_GET | 0x900;
+        } else {
+            if (giEntry.modIndex == 0) {
```
same magic number as above

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12715,30 +12738,69 @@ s32 func_8084DFF4(GlobalContext* globalCtx, Player* this) {
+                    }
+                    Audio_PlayFanfare(temp1);
+                }
+            } else {
```
probably else if here too

leggettc18 replied:
Addressed by 0c334615b65abe06a7d7e9a3ceb373c6e28e7203

briaguya0 reviewed (changes requested):
this is looking great! i'm super excited for how much easier this will make building things out!

leggettc18 commented:
To broadly address some of the repeated comments about the magic numbers and commented out stuff from the review, moving the `ModIndex` value to a more portable location is part of that planned relocation and cleanup I mentioned. As are the commented out stuff I need to delete, as well as the `GI_` and `ITEM_` values rando added that are no longer needed. Just wanted to make it clear those were being addressed without individually replying/reacting to each comment since there were a lot of them.

leggettc18 committed:
Merge branch 'develop-zhora' of https://github.com/HarbourMasters/Shipwright into get-item-rework

leggettc18 committed:
Removes unused function override

leggettc18 committed:
Replaces ::find with ::at and adds exception handling

leggettc18 committed:
Removes some commented out code.

leggettc18 committed:
Refactors rando's GetItemEntry array into two arrays.

One array is for the vanilla items that don't have GetItemEntries in
vanilla, the other is for rando exclusive items. They are stored in
separate arrays before getting added to the table so that we can apply
different modIndexes. The items in the first table have are handled
by the vanilla Item_Give, and the second table needed a custom
`Randomizer_Item_Give` function.

leggettc18 committed:
Renames, relocates, and implements ModIndex enum.

leggettc18 committed:
Removes now unused ItemIDs and GetItemIDs

Also makes all the necessary changes to other code that was still
using them indirectly through the GI to GID map that was removed.
There's quite a lot of changes here and I haven't had time to test them
yet.

leggettc18 commented:
The last batch of commits should resolve most of your comments @briaguya-ai. I think there's one or two still hanging. The last commit in particular though is quite large and I haven't had time to test all of it's changes yet. It *should* build but I have a feeling it broke a couple of things, haven't had time to buckle down and test it yet.

briaguya0 commented on soh/soh/Enhancements/custom-message/CustomMessageTypes.h:
```diff
@@ -27,7 +27,14 @@ typedef struct {
 
-#define GIMESSAGE_UNTRANSLATED(giid, iid, message) \
-    { giid, iid, message, message, message }
+GetItemMessage GIMESSAGE_UNTRANSLATED(u16 giid, u16 iid, std::string message) {
```
naming this in all caps seems kind of odd, but since we still have the `GIMESSAGE` macro maybe it keeps things consistent

leggettc18 replied:
It used to be a Macro (like GIMESSAGE above it still is) but I changed it to a function due to errors that started happening after I merged the custom message stuff into this get item rework branch. So it's named in all caps for backwards compatibility reasons.

leggettc18 replied:
Well I put it back as a macro and the errors I was having went away 🤷🏼

leggettc18 replied:
So unless you have any other comments about macro vs function this comment can be marked as resolved.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4738,31 +4558,166 @@ void Randomizer::CreateCustomMessages() {
+            default:
+                // We should never get here. If this branch of code executes,
+                // then you've added an item to extendedVanillaGetItemTable that
+                // should be in randoGetItemTable.
```
maybe we should throw an exception here?

leggettc18 replied:
Probably, would need to decide how to handle that exception exactly. Am I just throwing it for a crash handler to take care of or will the coding calling this function need to handle it?

briaguya0 replied:
i think crash handler, based on the comment it's something that we should never expect to happen outside of a dev making a mistake, so i don't see a need to handle it anywhere else

leggettc18 replied:
Addressed in 4b7548e99867d47a750c3916268d20050cde6067

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.h:
```diff
@@ -7,6 +7,8 @@
 #include <memory>
 #include <soh/Enhancements/randomizer/randomizerTypes.h>
 
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
```
i wonder if there's a better place for this to live since it's not rando specific

leggettc18 replied:
I think this is in a global area but it's in one of those header files that causes redefinition errors if rando imports it.

briaguya0 replied:
aha, it'd be nice to add a comment saying that (ideally we'd figure out the redefinition error but that feels out of scope for this PR)

leggettc18 replied:
I'll take another look and see if I can figure out something better before adding the comment.

leggettc18 replied:
Actually, I was wrong. It wasn't globally available anywhere. A search of the codebase shows it in `randomizer.h` (this one), `z_boss_va.c` (which I think was also me), and `imgui_impl_opengl3_loader.h`.

leggettc18 replied:
Addressed by e1084c0d6da6019f2cc29c57c8927dfb2b07e405

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1470,14 +1605,62 @@ extern "C" CustomMessageEntry Randomizer_GetHintFromCheck(RandomizerCheck check)
     return hintText;
 }
 
+extern "C" GetItemEntry ItemTable_Retrieve(int16_t getItemID) {
+    if (OTRGlobals::Instance->getItemModIndex != MOD_NONE) {
+        getItemID++; // counteracts the - 1 offset used for vanilla table
+    }
+
+    return ItemTableManager::Instance->RetrieveItemEntry(OTRGlobals::Instance->getItemModIndex, getItemID);
+}
+
+extern "C" GetItemEntry ItemTable_RetrieveEntry(s16 tableID, s16 getItemID) {
+    if (tableID == MOD_NONE) {
+        getItemID--; // counteracts the - 1 offset used for vanilla table
+    }
+    return ItemTableManager::Instance->RetrieveItemEntry(tableID, getItemID);
+}
```
it's not clear to me why we need to do the `--` when it's vanilla, and the `++` when it's not, mostly it's unclear how a `--` would counteract a `-1`

leggettc18 replied:
Oh I think that comments needs to be changed. Vanilla GetItemIds have GI_NONE as the first entry, but the first sGetItemTable entry corresponds to the second GI value, and all the code that was pulling from sGetItemTable was calling getItemId - 1. But the rando table doesn't have that problem so it doesn't need the - 1. The one doing ++ adds it back when it's a rando item. The other function is actually doing the opposite, subtracting one when it's _not_ a rando item. I'll fix the comment, or after some more cleanup it's possible we don't to adjust that value at all anymore, even for vanilla items.

leggettc18 replied:
Addressed by 69086f58c2b6bb00a273eefed4744ad456bba6ba

briaguya0 commented on soh/src/code/z_en_item00.c:
```diff
@@ -1208,38 +1208,56 @@ void EnItem00_Draw(Actor* thisx, GlobalContext* globalCtx) {
+        case MOD_NONE:
+            switch (giEntry.itemId) {
+                case ITEM_SONG_MINUET:
+                    color_slot = 0;
```
unrelated to this PR (mostly a note to self to make an issue for this) but it'd be good to not use magic numbers for the colors here

briaguya0 replied:
made an issue https://github.com/HarbourMasters/Shipwright/issues/1156

briaguya0 commented on soh/src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:
```diff
@@ -99,8 +99,10 @@ void ItemBHeart_Draw(Actor* thisx, GlobalContext* globalCtx) {
     if (gSaveContext.n64ddFlag) {
-        GetItem_Draw(globalCtx,
-                     Randomizer_GetItemModelFromId(Randomizer_GetRandomizedItemId(GI_HEART_CONTAINER_2, this->actor.id, this->actor.params, globalCtx->sceneNum)));
+        GetItem_Draw(globalCtx, ABS(Randomizer_GetRandomizedItem(GI_HEART_CONTAINER_2, this->actor.id,
+                                                                 this->actor.params, globalCtx->sceneNum)
+                                        .gi) -
+                                    1);
```
some funky formatting here

leggettc18 replied:
Yeah VSCode was doing it automatically. I guess I could probably just turn autoformat off for C/C++ files?

briaguya0 replied:
yeah, autoformat defaults can be funky sometimes, might be worth keeping it off until we have some sort of linting set up

i think the plan is to add linting to the ci

@louist103 and i were discussing linting the other day, the idea is to wait for a good time (like when we stop dev on rachael and start doing everything in zhora maybe) to add linting, then do a giant PR formatting everything, then enforce linting via ci for all PRs after that

leggettc18 replied:
Addressed by 4cc492f383094550d89b12c9759c83130d91f1e6, as well another spot or two that had similar weird formatting.

briaguya0 commented on soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c:
```diff
@@ -249,20 +247,17 @@ void ItemEtcetera_Draw(Actor* thisx, GlobalContext* globalCtx) {
-        if (randoGetItemId != GI_NONE) {
-            this->giDrawId = Randomizer_GetItemModelFromId(randoGetItemId);
+        if (randoGetItem.itemId != RG_NONE) {
+            this->giDrawId = ABS(randoGetItem.gi) - 1;
```
do we do this in other places? if so, could we macro it and add some comments next to the macro explaining why/how it works?

leggettc18 replied:
Yes it's to counteract how the `GID` gets turned into the `gi` value by the `GET_ITEM` macro. It multiplies it by either 1 or -1 according to the chest animation value and then adds 1. So this takes the absolute value and then subtracts 1. I can macro/function it or store the GID separately on the GetItemEntry, think I'd prefer the macro/function approach.

leggettc18 replied:
Addressed by 58c9f64a1689fbf3364beec0f362103b02c856a0

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -1594,9 +1408,14 @@ s32 func_808332E4(Player* this) {
-    GetItemEntry* giEntry = &sGetItemTable[this->getItemId - 1];
+    GetItemEntry giEntry;
+    if (this->getItemEntry.objectId == OBJECT_INVALID) {
+        giEntry = ItemTable_Retrieve(this->getItemId - 1);
```
would removing the `-1` here make it so less counteracting would be needed?

leggettc18 replied:
As mentioned in an earlier comment, the Vanilla GetItemTable's indices are off by one from GetItemId values. I can probably change that now that the Vanilla GetItemTable is a map though. Was trying not to change vanilla code too much if possible but that's probably what needs to happen.

leggettc18 replied:
Addressed by 69086f58c2b6bb00a273eefed4744ad456bba6ba

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -5042,8 +4861,12 @@ s32 func_8083B040(Player* this, GlobalContext* globalCtx) {
-                            giEntry = &sGetItemTable[D_80854528[sp2C] - 1];
-                            func_8083AE40(this, giEntry->objectId);
+                            if (this->getItemEntry.objectId == OBJECT_INVALID) {
+                                giEntry = ItemTable_Retrieve(D_80854528[sp2C] - 1);
```
another place that the `-1` is there, i'm still curious as to if removing that could clear up the logic or if it would make it more complex

leggettc18 replied:
Same as other comments, - 1 is from vanilla due to GetItemIds being off by one from corresponding entries in sGetItemTable.

leggettc18 replied:
Addressed by 69086f58c2b6bb00a273eefed4744ad456bba6ba

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -6324,38 +6152,43 @@ s32 func_8083E5A8(Player* this, GlobalContext* globalCtx) {
+                        ((Item_CheckObtainability(giEntry.itemId) != ITEM_NONE) && (giEntry.field & 0x20))) {
                         this->getItemId = -GI_RUPEE_BLUE;
-                        giEntry = &sGetItemTable[GI_RUPEE_BLUE - 1];
+                        giEntry = ItemTable_Retrieve(GI_RUPEE_BLUE - 1);
```
noting another `-1` in case that's a thing that we end up wanting to change

leggettc18 replied:
Addressed by 69086f58c2b6bb00a273eefed4744ad456bba6ba

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12703,16 +12536,20 @@ void func_8084DFAC(GlobalContext* globalCtx, Player* this) {
     if (this->unk_84F == 0) {
-        giEntry = &sGetItemTable[this->getItemId - 1];
+        if (this->getItemEntry.objectId == OBJECT_INVALID) {
+            giEntry = ItemTable_Retrieve(this->getItemId - 1);
```
noting another `-1` in case that's a thing that we end up wanting to change

leggettc18 replied:
Addressed by 69086f58c2b6bb00a273eefed4744ad456bba6ba

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -13316,14 +13192,14 @@ void func_8084F104(Player* this, GlobalContext* globalCtx) {
         }
         else {
-            GetItemEntry* giEntry = &sGetItemTable[D_80854528[this->exchangeItemId - 1] - 1];
+            GetItemEntry giEntry = ItemTable_Retrieve(D_80854528[this->exchangeItemId - 1] - 1);
```
i'm thinking replacing `D_80854528[this->exchangeItemId - 1]` is out of scope of this PR, but the other `-1` here fits the rest of the pattern

leggettc18 replied:
True, yeah haven't touched the exchange items at all, they all behave identically in vanilla and rando anyway.

leggettc18 replied:
Addressed by 69086f58c2b6bb00a273eefed4744ad456bba6ba

briaguya0 replied:
made an issue for the other part just in case we ever want to do some exchange item stuff in the future https://github.com/HarbourMasters/Shipwright/issues/1175

briaguya0 reviewed (commented):
overall looking really great!

marked some old comments as resolved, found a spot or two where formatting was odd, and left some new comments with questions

leggettc18 committed:
Re-implements GIMESSAGE_UNTRANSLATED as macro

leggettc18 committed:
Removes commented out function.

leggettc18 committed:
Throws exception if an invalid itemID is used

Addresses https://github.com/HarbourMasters/Shipwright/pull/1050#discussion_r943694857

leggettc18 committed:
Removes ARRAY_SIZE in favor of ARRAY_COUNT

ARRAY_COUNT already exists in `macros.h`, I just didn't find it before.

Addresses https://github.com/HarbourMasters/Shipwright/pull/1050#discussion_r943153833

leggettc18 committed:
Inverts CheckContainsRandoItem to CheckContainsVanillaItem.

Addresses https://github.com/HarbourMasters/Shipwright/pull/1050#discussion_r940895135

leggettc18 committed:
Cleanup, bugfixes, removing the `- 1`s from `z_player.c`

leggettc18 committed:
Fixes some funky formatting that got committed earlier.

leggettc18 committed:
Adds else if to added fanfare sound cases.

Addresses https://github.com/HarbourMasters/Shipwright/pull/1050#discussion_r940112924 and https://github.com/HarbourMasters/Shipwright/pull/1050#discussion_r940113492

leggettc18 committed:
Extends GetItemEntry to include getItemId

Also adapts some existing calls for both the entry and the id to only
get the entry.

leggettc18 committed:
Extends GetItemEntry to include GID.

This allows for using it later when drawing freestanding items.

Addresses https://github.com/HarbourMasters/Shipwright/pull/1050#discussion_r943168136

leggettc18 commented:
OK, that's all of Bria's comments addressed, next I'll branch off of this, merge in rando-next and PR that separately to rando-next so the #randomizer-playtesting folks can take a stab at it.

briaguya0 reviewed (approved):
this is so great!

:shipit:

leggettc18 committed:
Rando-specific items use new textId again.

This got lost when merging develop-zhora in because I didn't have custom
messages merged when I started this.

leggettc18 commented:
For the record, not confident to merge this just yet. Needs more playtesting and I'm still finding a few bugs. For example Shops were crashing because nothing was setting that global modIndex (already fixed, commit incoming for that), and there's some obscure bug I just found where I seem to only be able to pick up one Deku Stick per scene load from withered deku babas. After that the rest turn into Green Rupees.

leggettc18 committed:
Sets global modIndex to MOD_NONE on scene load

Fixes a crash when buying items in shops due to them
not triggering the rando code that normally sets these items.
May have also been crashing vanilla playthroughs.

leggettc18 committed:
Realized I had the bgm conditions wrong.

leggettc18 committed:
Fixes "static drops" (i.e. sticks from withered babas)

leggettc18 committed:
Fixes LACS/Prelude situation... again.

leggettc18 committed:
Fixes too many arguments error.

Not sure why this didn't fail to build on Windows before.

leggettc18 committed:
Fixes Link's Pocket items.

leggettc18 committed:
Simplifies sram init for rando-specific items

leggettc18 committed:
Fixes issues with approaching bottleable items.

leggettc18 committed:
Fixes Ruto's Letter.

It was accidentally getting classified as a rando item.

leggettc18 commented:
Alright this is getting some good playtesting in over on randomizer-playtesting, or more specifically the rando-next version of this PR. Some bugs still being found, will push fixes as I find them.

leggettc18 committed:
Should re-fix freestanding ice traps

leggettc18 committed:
Makes freestanding items set player->getItemEntry.

This prevents freestanding items from setting the global modIndex.
This is part of a larger transition that needs to happen to switch
to setting getItemEntries for all of the rando items. This prevents
some things that set getItemId of GI_MAX from granting a Fire Medallion
when the global modIndex is MOD_RANDOMIZER.

leggettc18 committed:
Makes sure we aren't using getItemEntry when not randoed.

garrettjoecox committed:
Replace Randomizer_GetRandomizedItemId with Randomizer_GetRandomizedItem and Randomizer_GetItemIdFromKnownCheck with Randomizer_GetItemFromKnownCheck

garrettjoecox commented on soh/src/code/z_play.c:
```diff
@@ -242,17 +242,19 @@ void GivePlayerRandoRewardZeldaLightArrowsGift(GlobalContext* globalCtx, Randomi
-        player->pendingFlag.flagID = 0x1E;
-        player->pendingFlag.flagType = FLAG_SCENE_TREASURE;
+        GetItemEntry getItem = Randomizer_GetItemFromKnownCheck(check, GI_ARROW_LIGHT);
+        if (player->pendingFlag.flagType == FLAG_NONE && GiveItemWithoutActor(globalCtx, getItem.getItemId)) {
```
Do we need to do this flag check and set via pending flags for the other GivePlayerRandoReward___?

leggettc18 replied:
Potentially, this was done specifically for this check as part of a fix for that specific check in the new system because it was also potentially competing with another check in the same area. The pending flags system in general was added in a previous PR.

leggettc18 replied:
What I probably want to do eventually is wrap that pendingFlag check and set in a function, but that's a bit out of scope for this PR I think.

garrettjoecox committed:
Introduce some new methods and migrate most actors to them

leggettc18 committed:
Fixes ocarina game skull kids to set player->getItemEntry

leggettc18 committed:
Sets `z_en_box.c` to set `player->getItemEntry`

garrettjoecox committed:
Fix logical errors and migrate most of the rest of the rando checks to GiveItemEntryFromActor

leggettc18 committed:
Merge branch 'get-item-rework-help1' of github.com:garrettjoecox/Shipwright into pr/garrettjoecox/2

garrettjoecox committed:
Use GiveItemEntryFromActorWithFixedRange in item00

leggettc18 committed:
Fixes Anju to set player->getItemEntry.

leggettc18 committed:
Merge branch 'get-item-rework-help1' of github.com:garrettjoecox/Shipwright into pr/garrettjoecox/2

garrettjoecox committed:
Add a few missing cases

leggettc18 committed:
Merge branch 'get-item-rework-help1' of github.com:garrettjoecox/Shipwright into pr/garrettjoecox/2

leggettc18 committed:
Merge pull request #2 from garrettjoecox/get-item-rework-help1

Replace Randomizer_GetRandomizedItemId with Randomizer_GetRandomizedI…

leggettc18 committed:
Additional fix for Skull Kid

leggettc18 committed:
Merge branch 'get-item-rework' of github.com:leggettc18/Shipwright into get-item-rework

leggettc18 committed:
Merge branch 'develop-zhora' of github.com:HarbourMasters/Shipwright into get-item-rework

leggettc18 committed:
Fixes vanilla ice traps and randomized ice smoke

leggettc18 commented:
Alright, thanks to playtesting done to #1191, we've identified several bugs and fixed the ones not specific to rando-next changes on this branch first. So far everything seems to work fine on both Vanilla and Randomizer. @briaguya-ai there's been quite a few changes since you last reviewed if you want to take another review or either this or the rando-next version! Unless any more bugs are found, this should be ready for review and merge. Alternatively we can merge the rando-next version first and if any more bugs are found I'll continue to apply any applicable fixes here first.

leggettc18 committed:
Fixes rendering of treasure chest game items.

briaguya0 commented on soh/include/functions.h:
```diff
@@ -453,7 +454,9 @@ u32 Actor_TextboxIsClosing(Actor* actor, GlobalContext* globalCtx);
 void Actor_GetScreenPos(GlobalContext* globalCtx, Actor* actor, s16* x, s16* y);
 u32 Actor_HasParent(Actor* actor, GlobalContext* globalCtx);
-s32 GiveItemWithoutActor(GlobalContext* globalCtx, s32 getItemId);
+s32 GiveItemEntryWithoutActor(GlobalContext* globalCtx, GetItemEntry getItemEntry);
+s32 GiveItemEntryFromActor(Actor* actor, GlobalContext* globalCtx, GetItemEntry getItemEntry, f32 xzRange, f32 yRange);
+void GiveItemEntryFromActorWithFixedRange(Actor* actor, GlobalContext* globalCtx, GetItemEntry getItemEntry);
```
i think these are new, not sure about the naming `GiveItemEntry`, we're still just giving items, but passing in an `GetItemEntry` instead of an id. I think just using `GiveItemWithoutActor`/`GiveItemFromActor`/`GiveItemFromActorWithFixedRange` should work

also, not clear why the fixed range one would be a void

leggettc18 replied:
They are new, and @garrettjoecox actually wrote those. The naming I could go back and forth on, and as while we did replace the `GiveItemWithoutActor` that already existed and thus could have used that name, the existing functions equivalent to `GiveItemEntryFromActor` and `GiveItemEntryFromActorWithFixedRange` do still exist and are still used in vanilla code. So if we remove `Entry` from these function names, what will we call those existing functions when it's time to document them.

`GiveItemEntryFromActorWithFixedRange` probably returns a void because `func_8002F554`, it's already existing counterpart used in vanilla that does the same thing except only sets `player->getItemId`, also returns void. No reason it couldn't return `s32` if desired.

briaguya0 replied:
I just feel like the current naming implies we're *giving* an *itementry*, when we're actually giving an item based on a *GetItemEntry*

leggettc18 replied:
I mean sort of? These functions don't directly give the item anyway, and the old getItemId based ones didn't either, they all just set a few values on the player struct that get read later to do the actual item giving.

briaguya0 replied:
hmm, yeah, coming up with good names for these ones is tough

only names that comes to mind now are
`GiveItemEntryWithoutActor` -> `SetPlayerGetItemDataFromEntry` 
`GiveItemEntryFromActor` -> `SetPlayerGetItemDataFromEntryWithActor`
`GiveItemEntryFromActorWithFixedRange` -> `SetPlayerGetItemDataFromEntryWithFixedRangeActor`

and then the old func named ones could change to
`SetPlayerGetItemDataFromId` 
`SetPlayerGetItemDataFromIdWithActor`
`SetPlayerGetItemDataFromIdWithFixedRangeActor`

but those names are long....

leggettc18 replied:
Other places in the code often shortened `GetItemID` and `GetItemEntry` variable names into `giid` and `giEntry`, so maybe it wouldn't be too out of place to call these `SetPlayerGIEntry/SetPlayerGIID` etc.? I could also add some more descriptive comments and call it a day while we await a more thorough cleanup.

briaguya0 replied:
not the biggest fan of SetPlayerGIEntry because we're setting both the entry and id in that

I think comments for now seem like the best call. That combined with an issue for renaming `func_8002F434` etc. with a note saying we should rename the `GiveItemEntry`ones to be consistent with whatever naming scheme we come up with would be perfect imo

briaguya0 commented on soh/soh/Enhancements/custom-message/CustomMessageManager.cpp:
```diff
@@ -46,7 +47,7 @@ void CustomMessageManager::ReplaceColors(std::string& string) {
 }
 
-void CustomMessageManager::FormatCustomMessage(std::string& message, ItemID iid) {
+void CustomMessageManager::FormatCustomMessage(std::string& message, uint16_t iid) {
```
it would be nice if we could use types that mean something instead of just a `uint16_t` in places like this, it's definitely not something i'd want to prevent merging

i kind of want to make a code cleanup issue for it but if it's the best solution to the issues you ran into then it's probably not worth making one

leggettc18 replied:
I think this one could actually have `ItemID`, I don't think I ever send `RG_` values to this one, but there are some other functions that were taking or a returning a `GetItemID`, but that now might need to take a `RandomizerGet` value as well, and that causes errors unless I add a typecast which felt worse to me. Not to mention mods with custom items that need text and won't necessarily have a corresponding enum to use for these IDs. This one in particular I probably changed for consistency with the other ones that I mentioned. Maybe I could `#define GetItemID uint16_t` or something like that?

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2365,11 +2138,138 @@ GetItemID Randomizer::GetItemFromGet(RandomizerGet randoGet, GetItemID ogItemId)
+        case RG_BUY_BOMBS_30:
+        case RG_BUY_BOMBS_535:
+        case RG_BUY_RED_POTION_40:
+        case RG_BUY_RED_POTION_50:
```
come to think of it this is a much cleaner alternative to `x == y ||` than using an array like we discussed here: https://github.com/HarbourMasters/Shipwright/pull/1249/files#r949156153

briaguya0 commented on soh/soh/OTRGlobals.h:
```diff
@@ -95,13 +98,14 @@ u8 Randomizer_GetSettingValue(RandomizerSettingKey randoSettingKey);
-s32 Randomizer_GetItemIDFromGetItemID(s32 getItemId);
-s32 Randomizer_GetRandomizedItemId(GetItemID ogId, s16 actorId, s16 actorParams, s16 sceneNum);
-s32 Randomizer_GetItemIdFromKnownCheck(RandomizerCheck randomizerCheck, GetItemID ogId);
+s32 Randomizer_GetItemIDFromGetItemID(s32 getItemId); // TODO: Remove me
```
todo?

leggettc18 replied:
Ah, forgot about that one, need to double check if that's still used anywhere.

leggettc18 replied:
Still being used in `z_sram.c`, but I think I can actually make it so it's not used anymore.

leggettc18 replied:
Resolved by e1b403a31b1dde6ae190a2a8352d1f0fd69073dd

briaguya0 commented on soh/src/code/z_actor.c:
```diff
@@ -1973,12 +1975,44 @@ s32 GiveItemWithoutActor(GlobalContext* globalCtx, s32 getItemId) {
+    GiveItemEntryFromActor(actor, globalCtx, getItemEntry, 50.0f, 10.0f);
+}
+
+// TODO: Rename to GiveItemIdFromActor or similar
```
just a note, as of writing, it is still named `func_8002F434` in https://github.com/zeldaret/oot/blob/master/src/code/z_actor.c

leggettc18 replied:
Yes, that's part of why we didn't go ahead and change it. This comment just being here improves search-ability and helps clear up what that function does even without the rename. Also since decomp moves kinda slow I thought eventually we might want to document some things ourselves down the line.

briaguya0 commented on soh/src/code/z_actor.c:
```diff
@@ -1997,6 +2031,8 @@ s32 func_8002F434(Actor* actor, GlobalContext* globalCtx, s32 getItemId, f32 xzR
     return false;
 }
 
+// TODO: Rename to GiveItemIdFromActorWithFixedRange or similar
```
just a note, as of writing, it is still named `func_8002F554` in https://github.com/zeldaret/oot/blob/master/src/code/z_actor.c

briaguya0 commented on soh/src/code/z_en_item00.c:
```diff
@@ -509,8 +510,13 @@ void EnItem00_Init(Actor* thisx, GlobalContext* globalCtx) {
-        func_8002F554(&this->actor, globalCtx, getItemId);
+        getItem = Randomizer_GetRandomizedItem(getItemId, this->actor.id, this->ogParams, globalCtx->sceneNum);
+        getItemId = getItem.getItemId;
+        if (!gSaveContext.n64ddFlag) {
+            func_8002F554(&this->actor, globalCtx, getItemId);
+        } else {
```
should we have this block here? seems it might make sense to move it up a couple lines outside the `if ((gSaveContext.n64ddFlag || getItemId != GI_NONE) && !Actor_HasParent(&this->actor, globalCtx)) {` block, or maybe rework it to be something like

```c
if (!Actor_HasParent(&this->actor, globalCtx)) {
    if (!gSaveContext.n64ddFlag) {
        if (getItemId != GI_NONE) {
            func_8002F554(&this->actor, globalCtx, getItemId);
        }
    } else {
        GiveItemEntryFromActorWithFixedRange(&this->actor, globalCtx, Randomizer_GetRandomizedItem(getItemId, this->actor.id, this->ogParams, globalCtx->sceneNum));
    }
} 
```

mostly just trying to think of ways to avoid messing with the local `getItemId` var when rando'd because we're not using it (i know the nested ifs in the example code there aren't the best

leggettc18 replied:
Hmm, I'll revisit this and see if we can clean it up. Worth noting is that, with few exceptions, we actually set both `getItemEntry` and `getItemId` In our `GiveItemEntry...` functions. This is to prevent odd scenarios with things like bugs, fish, and fairies, that actually set `GI_MAX` as the player's `getItemId`. So if you're picking up an item nearby these, you don't have your `getItemEntry` being whatever's in the item and your `getItemId` being `GI_MAX`, since the item will supercede the `GI_MAX` from the "critters".

leggettc18 replied:
Yeah I see what you mean now, yeah this can definitely be cleaned up.

leggettc18 replied:
Resolved by e5bca727ac2bd799e6354b1052e188a29451f365

briaguya0 replied:
Looks like with the latest commit we won't hit the else if we're rando'd and getItemId is GI_NONE. 

That is slightly different logically, is that OK?

leggettc18 replied:
Ah good catch the semantics are a bit different on this one. If the getItemId were `GI_NONE` when not rando'd we would enter the else block when we probably shouldn't be doing anything at all. Was trying to remove a level of nesting from the if but in hindsight definitely not the correct play. I'll fix that.

briaguya0 commented on soh/src/code/z_en_item00.c:
```diff
@@ -666,10 +672,15 @@ void func_8001E304(EnItem00* this, GlobalContext* globalCtx) {
         if (!Actor_HasParent(&this->actor, globalCtx)) {
-            func_8002F434(&this->actor, globalCtx, this->getItemId, 50.0f, 80.0f);
+            if (!gSaveContext.n64ddFlag || getItemEntry.getItemId == GI_NONE) {
+                func_8002F434(&this->actor, globalCtx, getItemEntry.getItemId, 50.0f, 80.0f);
```
i guess i'm just not a fan of the pattern of calling `GetRandomizedItem` when we aren't rando'd, nothing against it logically, just name wise it feels odd

leggettc18 replied:
Yeah this can definitely be cleaned up a bit. I'm not actually sure why we get a pointer to the Player here either, it doesn't seem to be used anywhere.

leggettc18 replied:
Resolved by 2bbd618599d9bd96250798daffbb7fbb1677db60

briaguya0 reviewed (approved):
this is looking great! left a couple comments (mostly naming related)

leggettc18 committed:
Removes unused `Randomizer_GetItemIdFromGetItemId`.

leggettc18 committed:
Cleans up an if statement for item00.

leggettc18 committed:
Cleans up another if statement in item00

leggettc18 committed:
This should fix a bug with the Gerudo Archery minigame.

I wasn't able to get the bug to happen after making this change.

leggettc18 committed:
Merge remote-tracking branch 'origin' into get-item-rework

leggettc18 committed:
Documents our new GiveItemEntry fuctions.

leggettc18 committed:
Uses more descriptive type name for ItemIDs for creating custom messages.

leggettc18 committed:
Fixes potential issue with if statement.

leggettc18 committed:
Fixes missed type change.

leggettc18 commented:
All review comments that are in scope should be addressed at this point. Also @briaguya-ai after this gets merged, you can merge #1191 into rando-next. I've been periodically merging this PR into that PR so it has everything from here plus everything needed to make features only in rando-next work with it. It also already handled a nasty merge conflict that popped up in the ImGui section of `randomizer.cpp`

briaguya0 commented:
> after this gets merged, you can merge https://github.com/HarbourMasters/Shipwright/pull/1191 into rando-next.

yeah, my merging plan is:
* wait for this to get into zhora
* merge https://github.com/HarbourMasters/Shipwright/pull/1191 into rando-next
* merge zhora into rando-next (probably will already be up to date because of the merging you've been doing into the rando-next version of this, but it's possible a few other things will have made it into zhora by then)
